community/kernel/src/main/java/org/neo4j/unsafe/impl/batchimport/RelationshipGroupCache.java

```diff
@@ -0,0 +1,269 @@
+        this.highNodeId = highNodeId;
+
+        long memoryDedicatedToCounting = 2*highNodeId;
+        long memoryLeftForGroupCache = maxMemory - memoryDedicatedToCounting;
+        if ( memoryLeftForGroupCache < 0 )
+        {
+            throw new IllegalArgumentException( "Too little memory to cache any groups, provided " +
+                    bytes( maxMemory ) + " where " + bytes( memoryDedicatedToCounting ) +
+                    " was dedicated to group counting" );
+        }
+        this.cache = arrayFactory.newByteArray( memoryLeftForGroupCache / GROUP_ENTRY_SIZE, new byte[GROUP_ENTRY_SIZE] );
+    }
+
+    public void incrementGroupCount( long nodeId )
+    {
+        int count = groupCountCache.getShort( nodeId, 0 ) & 0xFFFF;
+        count++;
+        if ( (count & ~0xFFFF) != 0 )
+        {
+            throw new IllegalStateException(
+                    "Invalid number of relationship groups for node " + nodeId + " " + count );
+        }
+        groupCountCache.setShort( nodeId, 0, (short) count );
+    }
+
+    /**
+     * Getter here because we can use this already allocated data structure for other things in and
+     * around places where this group cache is used.
+     */
+    ByteArray getGroupCountCache()
+    {
+        return groupCountCache;
+    }
+
+    /**
+     * @param fromNodeId inclusive
+     * @return toNodeId exclusive
+     */
+    public long prepare( long fromNodeId )
+    {
+        cache.clear(); // this will have all the "first" bytes set to 0, which means !inUse
+        this.fromNodeId = fromNodeId; // keep for use in put later on
+
+        highCacheId = 0;
+        for ( long nodeId = fromNodeId; nodeId < highNodeId; nodeId++ )
+        {
+            short count = groupCountCache.getShort( nodeId, 0 );
+            if ( highCacheId + count > cache.length() )
+            {
+                // Cannot include this one, so up until the previous is good
+                return this.toNodeId = nodeId;
+            }
+            offsets.set( rebase( nodeId ), highCacheId );
+            highCacheId += count;
+        }
+        return this.toNodeId = highNodeId;
+    }
+
+    private long rebase( long toNodeId )
+    {
+        return toNodeId - fromNodeId;
+    }
+
+    public boolean put( RelationshipGroupRecord groupRecord )
+    {
+        long nodeId = groupRecord.getOwningNode();
+        assert nodeId < highNodeId;
+        if ( nodeId < fromNodeId || nodeId >= toNodeId )
+        {
+            return false;
+        }
+
+        long baseIndex = offsets.get( rebase( nodeId ) );
+        // grouCount is extra validation, really
+        int groupCount = groupCountCache.getShort( nodeId, 0 );
+        long index = scanForFreeFrom( baseIndex, groupCount, groupRecord.getType() );
+
+        // Put the group at this index
+        cache.setByte( index, 0, (byte) 1 );
+        cache.setShort( index, 1, (short) groupRecord.getType() );
```

Comment:
Wait, `RELATIONSHIP_GROUP_MAXIMUM_ID_BITS` in `StandardFormatSettings` is 35, how that will work?